### PR TITLE
LPD-5729 Autogenerate Headless API client code for Playwright

### PR DESCRIPTION
**Pull request by 4lejandrito**
Hi!

https://liferay.atlassian.net/browse/LPS-206277.

This PR generates automatic Headless client TypeScript code to use in the playwright tests.

It uses our global OpenAPI endpoint and [openapi-typescript-codegen](https://github.com/ferdikoomen/openapi-typescript-codegen):

```bash
rm -rf headless && curl 'http://localhost:8080/o/openapi/openapi.yaml' -u test@liferay.com:test > openapi.yaml && npx openapi-typescript-codegen --input openapi.yaml -o ./headless && rm -rf openapi.yaml && npm run format
```

### CODE REVIEW

**Comment by izaera** [comment]
Maybe we can add the command `rm -rf headless && curl 'http://localhost:8080/o/openapi/openapi.yaml' -u test@liferay.com:test > openapi.yaml && npx openapi-typescript-codegen --input openapi.yaml -o ./headless && rm -rf openapi.yaml && npm run format` to a `package.json` script? It looks useful for everybody... :thinking:

**Comment by izaera** [comment]
LGTM!

I still don't get the difference between `helpers` and `utils`, though, but I will ask in a thread in slack what people think...

**Comment by izaera** [comment]
ci:test:playwright

**Comment by izaera** [comment]
Mmm, related to this conversation -> https://liferay.slack.com/archives/C06A59K5D46/p1705411600313329, why did you decide to put things in a `headless` folder?

I understand it is part of the core test infrastructure, but at the same time I'm sure other teams may provide more shared utilities similar to this (especially us, frontend infra) so, how will we avoid having a lot of folders that pollute the `playwright` folder namespace? Ideally we wouldn't want to have a myriad folders at the top level, right...

**Comment by 4lejandrito** [comment]
I picked headless for no good reason. I think we can move it to wherever feels best.

**Comment by 4lejandrito** [comment]
> Maybe we can add the command `rm -rf headless && curl 'http://localhost:8080/o/openapi/openapi.yaml' -u test@liferay.com:test > openapi.yaml && npx openapi-typescript-codegen --input openapi.yaml -o ./headless && rm -rf openapi.yaml && npm run format` to a `package.json` script? It looks useful for everybody... 🤔

This could be added as an npm script. It is only meant to be used when adding a new API or method. And it should warn you that you need to restore the credentials in the OpenAPI file.

**Comment by izaera** [comment]
>  It is only meant to be used when adding a new API or method.

Would that be the headless team only or everybody? I guess headless team only, right? In that case, do as you wish (add it or keep it internal, I'd say :shrug: ). If we end up having too much scripts, we can decide how to give it some order later....

> I picked headless for no good reason. I think we can move it to wherever feels best.

"wherever"... :roll_eyes:  Do you have any proposal? :sweat_smile:  Inside `util` for now? :shrug:  Until we think of a better place...

**Comment by 4lejandrito** [comment]
1. We as headless can take care of it for now. I will describe it in the docs.
2. Util sounds good, but in a separate folder. Do we want to mark the folder as autogenerated somehow? like `util/autogenerated/headless`. It is horrible, but clear :laughing: .

**Comment by izaera** [comment]
> Do we want to mark the folder as autogenerated somehow?

I'd say no because it's just you (headless team) who handle the autogeneration. We (the rest) "don't care" about how the files got there as long as they can be used to access our headless APIs... :shrug:

**Comment by izaera** [comment]
Mmm, apparently the PR broke Playwright tests :grimacing:

**Comment by 4lejandrito** [comment]
I'll look into it.

**Comment by 4lejandrito** [comment]
I think https://github.com/liferay-frontend/liferay-portal/pull/3936/commits/4d3eaef2bf2731bc74e5fefa275130074e86c1e2 will fix it.

**Comment by 4lejandrito** [comment]
ci:test:playwright

**Comment by 4lejandrito** [comment]
ci:test:sf

**Comment by 4lejandrito** [comment]
Failure no longer related :partying_face:

File: modules/test/playwright/headless/core/OpenAPI.ts [hunk 1]
```diff
@@ -21,7 +21,7 @@ export type OpenAPIConfig = {
 };
 
 export const OpenAPI: OpenAPIConfig = {
-	BASE: 'http://localhost:8080/o',
+	BASE: liferayConfig.environment.baseUrl + '/o',
```
Comment (izaera) [inline modules/test/playwright/headless/core/OpenAPI.ts:24]: :open_mouth: :+1:

**Comment by izaera** [comment]
Looks good :+1: 

However, two things:

1. Didn't we say we were going to move the `headless` top folder somewhere? `utils/headless` for example? :thinking: 
2. Rebase error due to the formatting PR that was sent yesterday, most probably (I had the same problem with the login fixture PR) :-/

**Comment by 4lejandrito** [comment]
> Looks good 👍
> 
> However, two things:
> 
>     1. Didn't we say we were going to move the `headless` top folder somewhere? `utils/headless` for example? 🤔
> 
>     2. Rebase error due to the formatting PR that was sent yesterday, most probably (I had the same problem with the login fixture PR) :-/

Will fix it today.

**Comment by 4lejandrito** [comment]
Just push forced. Moved to `utils/headless` and created an npm script to run the regeneration when needed.

**Comment by 4lejandrito** [comment]
ci:test:playwright

**Comment by 4lejandrito** [comment]
ci:test:playwright

**Comment by 4lejandrito** [comment]
ci:test:playwright

**Comment by izaera** [comment]
Very good :ok_hand: 

There are more conflicts today... One strategy I'm going to test to avoid conflicts when touching too many files is to send PRs in parts: 

1. one part adds the functionality but doesn't use it anywhere (eg: login fixture, headless helpers)
2. then either:
    1. send PRs to the teams changing their code to use the new functionality
    2. send a sample PR then let teams change their code as they see it fits their schedules

In case you want to consider it for this PR.

**Comment by 4lejandrito** [comment]
ci:test:playwright

**Comment by 4lejandrito** [comment]
ci:test:playwright

**Comment by izaera** [comment]
The `export-import` test is a bit flaky from what I've seen in many runs (local and CI) so I guess the PR is safe...

**Comment by izaera** [comment]
ci:test:sf

**Comment by izaera** [comment]
LGTM, you can forward when you like @4lejandrito 

However you may need to split it in two parts as I said yesterday (one to add the helpers only that can be merged easily without conflicts), another one to make the tests use the helpers (this one will conflict a lot of times, for sure :grimacing: ).

Alternatively, we can find a way to mark the helpers as deprecated and announce how they should be replaced in the deprecation notice and in slack public channel. What do you think? That's what I plan to do with the login fixture...

**Comment by izaera** [comment]
Mmm, we may need to change `tsc` target in `tsconfig.json` in this PR to fix the SF error...

Since we are running tests in Node, this can be used to find the correct settings -> https://github.com/microsoft/TypeScript/wiki/Node-Target-Mapping

We are using node 16 currently, in DXP.

@4lejandrito Can you add a commit to this PR to fix that or do you prefer that I send a PR alone to account for it?

**Comment by 4lejandrito** [comment]
Hey @izaera I am happy to do that, but we identified issues in fedora with this PR. When we fix them we will send everything.

**Comment by 4lejandrito** [comment]
ci:test:playwright

**Comment by 4lejandrito** [comment]
ci:test:playwright

**Comment by 4lejandrito** [comment]
Last test run passed :partying_face: 

@jaimelr10 please take it from here, resolve conflicts and investigate the failures on fedora.

**Comment by 4lejandrito** [comment]
ci:test:sf

**Comment by 4lejandrito** [comment]
ci:test:playwright

**Comment by 4lejandrito** [comment]
ci:test:sf

**Comment by 4lejandrito** [comment]
ci:test:playwright

**Comment by 4lejandrito** [comment]
ci:forward

**Pull request closed without merging**